### ISSUE TITLE
Unable to compile gridlabd with clang on Mac

### ISSUE THREAD

**Issue by kdheepak** [gridlab-d/gridlab-d#1080]
- [x] I have not found an open issue describing the same problem.
- [x] I went through "CONTRIBUTING.MD" before deciding to submit the issue.
- [x] While building GridLAB-D, I followed all steps as detailed in GridLAB-D Wiki.

Has anyone else been able to do this with the latest version of clang? There appears to be some function calls to `fgets` that are not compatible with clang.

**Comment by ftuffner** [comment]
We haven't looked into it thoroughly, but we do have one GridLAB-D user who has been trying to use the latest version of clang in MacOS to compile GridLAB-D.  This required a couple code changes (warnings that are now errors), including an `fgets` error inside csvreader.cpp (as I recall).  These changes have not been pushed to the repository yet, as we're still encountering problems.

With the code fixes, GridLAB-D compiles, but some autotests inside the powerflow module associated with step-up delta-GWye transformers are failing.  We're still looking into this to see why.

**Comment by kdheepak** [comment]
I remember making some changes to get it to work with clang. I can submit that as a PR. But I don't know why these changes are required, I recursively went through and fixed issues as I had them.

**Comment by kdheepak** [comment]
These changes are located in a fork [here](https://github.com/kdheepak/gridlab-d/commit/d03a8a1f022bc9a3585f094b224da354fbd61dd1).

**Comment by xzhu1991** [comment]
This is the error I got while I was trying to compile GridLAB-D:
climate/csv_reader.cpp:108:34: error: ordered comparison between pointer and zero ('char *' and 'int')
        while(fgets(line, 1024, infile) > 0){

**Comment by trevorhardy** [comment]
What version of clang are you using, specifically? I'm still on macOS 10.12 (Sierra) and haven't had any problems with clang.

As a temporary work-around, you should be able to get it to compile with gcc/g++. I've been moving back and forth between gcc-6 and clang 8.0.0 depending on the needs of the project.  You can install gcc via [Homebrew](https://brew.sh) very easily and then reference it when running `./configure` such as:

`./configure CC=/usr/local/bin/gcc-6 CXX=/urs/local/bin/g++-6`

**Comment by kdheepak** [comment]
Thanks for the comment @trevorhardy. I'll confirm which version we are using in a bit, I'm on my mobile at the moment. `gcc` would probably work and we can test it and get back to you, however the issue is that we have shared libraries that are built using clang that we would like to integrate with GridLAB-D. To my understanding, this would not be possible, or at least not recommended

**Comment by kdheepak** [comment]
```bash 
clang --version

Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by kdheepak** [comment]
Hmm. I'm not able to get this to work with gcc either.

```bash
...
configure: error: in `/Users/$USER/GitRepos/gridlab-d':
configure: error: C++ preprocessor "/lib/cpp" fails sanity check
See `config.log' for more details
```

[config.log](https://gist.github.com/kdheepak/6a3204ce817cea1f85ed22e41a1e896a)

**Comment by afisher1** [comment]
Hey @kdheepak,

Another college posted some fixes to develop so that he could compile using clang on his mac. He has Sierra and is able to compile as well as have all the autotests passing. I have another person here verifying that this is fixed but I'd also like you to check as well.

**Comment by kdheepak** [comment]
Thanks @afisher1, I'm now able to install gridlabd on osx with the latest clang.

**Comment by kdheepak** [comment]
@dchassin I was able to build from `develop` branch successfully.

**Comment by kdheepak** [comment]
Multiple `autotests` fail even if I'm able to build gridlabd successfully. Also, builds with gcc do not work with `develop`. I'm using gcc 7.2.0 installed with homebrew.

```bash
$ /usr/local/Cellar/gcc/7.2.0/bin/c++-7 --version
c++-7 (Homebrew GCC 7.2.0) 7.2.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by nightlark** [comment]
It seems like building gridlab-d with clang on Mac are broken again -- is there some form of CI setup that tests builds on different platforms regularly?

**Comment by afisher1** [comment]
Hi @nightlark,

We are currently only supporting gcc builds on all platforms. We currently use buildbot for our automated build process on in-house VMs.

Andy

**Comment by eberleim** [comment]
Fixed by PR #1236
